### DISCUSSION TITLE
ThemeManager: replaced methods

### DISCUSSION THREAD

**Discussion by Michi-S7000** [ControlzEx/ControlzEx#199] [Q&A] [answered]
Hello
In our application we used until now MahApps.Metro Version 1.6.5.1. and the following lines of code:
var themeRed = ThemeManager.DetectAppStyle(Application.Current);
ThemeManager.ChangeAppStyle(Application.Current, ThemeManager.GetAccent("s7000Red"), themeRed.Item1);

(Like discribed in the guides)

Now, the ThemeManager is in the ContolzEx NuGet Package and the methods DetectAppStyle(...) and ChangeAppStyle(...) no longer exist.
I tried to replace with DetectTheme(Application.Current.Resources) and ChangeTheme(...), but had no success. The method DetectTheme returns null, even while the ResourceDictionary passed in is the same as in the old version.

Can someone help me to get running this peace of code again?
Thanks and best regards, Michael

**Comment by batzen** [comment]
How do you set the initial app theme?
The markers for detecting themes are important.
When you say the ResourceDictionary passed in is the same, what does that mean exactly? Do you have a custom theme stored there?

**Reply by Michi-S7000** [reply]
In the old version it was this way, in the OnStartup-method of App.xaml.cs:
ThemeManager.AddAccent("s7000Blue", new Uri("pack://application:,,,/ABS7000;component/Resources/s7000AccentBlue.xaml"));
ThemeManager.AddAccent("s7000Red", new Uri("pack://application:,,,/ABS7000;component/Resources/s7000AccentRed.xaml"));

In the new version I tried:
Theme themeBlue = new Theme(new LibraryTheme(new Uri("pack://application:,,,/ABS7000;component/Resources/s7000AccentBlue.xaml"), MahAppsLibraryThemeProvider.DefaultInstance));
ThemeManager.Current.AddTheme(themeBlue);

Theme themeRed = new Theme(new LibraryTheme(new Uri("pack://application:,,,/ABS7000;component/Resources/s7000AccentRed.xaml"), MahAppsLibraryThemeProvider.DefaultInstance));
ThemeManager.Current.AddTheme(themeRed);

That's different then before.

**Reply by batzen** [reply]
Could you provide a small repro application?
That would be way easier to analyze.

**Reply by Michi-S7000** [reply]
Because it's a huge amount of work to provide a repro app, I try to make a small sample project to learn about ThemeManager. But I'm facing already in the first lines of code some problems. Is the ThemeManager compatible with .net 8 apps?

**Reply by batzen** [reply]
It's compatible with all .net versions greater than or equal to.net 6 and with all .net framework versions greater than or equal to 4.6.2.

**Reply by Michi-S7000** [reply]
Thank you! Now I'm one step further. But now I tried to figure out how SyncTheme works. I would expect that the theme set in the MainWindow would be transferred to other windows. Can you explain to me how this works?

**Reply by batzen** [reply]
You either have to use `ThemeManager.Current.ThemeSyncMode = ThemeSyncMode.SyncWithAppMode;` (you can choose whichever `ThemeSyncMode` you like of course) before calling `SyncTheme` or you have to pass a sync mode to `SyncTheme`.

**Reply by Michi-S7000** [reply]
Thank you. But It's not working as expected yet. I have the following Code:

`public partial class MainWindow : MetroWindow
{
    public MainWindow()
    {
        InitializeComponent();
        ThemeManager.Current.ThemeSyncMode = ThemeSyncMode.SyncWithAppMode;
        ThemeManager.Current.ChangeTheme(this, "Light.Blue");
        ThemeManager.Current.SyncTheme();
    }

    private void Button_Click(object sender, RoutedEventArgs e)
    {
        new SecondWindow().Show();
    }

    private void ChangeTheme_Click(object sender, RoutedEventArgs e)
    {
        ThemeManager.Current.ChangeTheme(this, "Dark.Blue");
        ThemeManager.Current.SyncTheme();
    }
}`
and 

` public partial class SecondWindow : MetroWindow
 {
     public SecondWindow()
     {
         ThemeManager.Current.SyncTheme();
         InitializeComponent();
         
     }

     private void Button_Click(object sender, RoutedEventArgs e)
     {
         ThemeManager.Current.ChangeTheme(this, "Dark.Blue");
         ThemeManager.Current.SyncTheme();
     }
 }`

When I run the application, the MainWindow appears in Light.Blue, as expected. When I click on the Button to show the second window, it appears totaly black. I can not see the buttons on this window. I don't understand why?

**Reply by Michi-S7000** [reply]
I expect that the SecondWindow would also appear in Light.Blue. The ChangeTheme-Buttons on the windows work as expected. The Theme is changing.

**Reply by batzen** [reply, marked as answer]
SynTheme is meant to sync the current application theme with windows settings (accent color, app mode (dark/light).

Your code only sets the theme on "this", so on one specific window. If you want to change the application theme, for all windows, you have to pass Application.Current instead of a specific window.

**Reply by Michi-S7000** [reply]
Thank you for the quick answer. This did the trick. I also removed all calls of SyncTheme(). Apparently it works without it. Why is this method needed?

**Reply by Michi-S7000** [reply]
Sorry, I see the answer is in your first sentence.
